pysph/sph/wc/kernel_correction.py

```diff
@@ -0,0 +1,198 @@
+from pysph.sph.equation import Equation
+from pysph.sph.wc.density_correction import gj_solve
+
+
+class KernelCorrection(Equation):
+    r"""**Kernel Correction**
+
+    .. math::
+            \mathbf{f}_{a} = \frac{\sum_{b}\frac{m_{b}}{\rho_{b}}
+            \mathbf{f}_{b}W_{ab}}{\sum_{b}\frac{m_{b}}{\rho_{b}}W_{ab}}
+    References
+    ----------
+    .. [Bonet and Lok, 1999] Bonet, J. and Lok T.-S.L. (1999)
+        Variational and Momentum Preservation Aspects of Smoothed
+        Particle Hydrodynamic Formulations.
+
+    """
+
+    def initialize(self, d_idx, d_cwij):
+        d_cwij[d_idx] = 0.0
+
+    def loop(self, d_idx, s_idx, d_cwij, s_m, s_rho, WIJ):
+        d_cwij[d_idx] += s_m[s_idx] * WIJ / s_rho[s_idx]
+
+
+class GradientCorrectionPreStep(Equation):
+
+    def __init__(self, dest, sources, dim=2):
+        self.dim = dim
+        super(GradientCorrectionPreStep, self).__init__(dest, sources)
+
+    def initialize(self, d_idx, d_m_mat):
+        i = declare('int')
+        for i in range(9):
+            d_m_mat[9 * d_idx + i] = 0.0
+
+    def loop_all(self, d_idx, d_m_mat, s_m, s_rho, d_x, d_y, d_z, d_h, s_x,
+                 s_y, s_z, s_h, KERNEL, NBRS, N_NBRS):
+        x = d_x[d_idx]
+        y = d_y[d_idx]
+        z = d_z[d_idx]
+        h = d_h[d_idx]
+        i, j, s_idx, n = declare('int', 4)
+        xij = declare('matrix(3)')
+        dwij = declare('matrix(3)')
+        n = self.dim
+        for k in range(N_NBRS):
+            s_idx = NBRS[k]
+            xij[0] = x - s_x[s_idx]
+            xij[1] = y - s_y[s_idx]
+            xij[2] = z - s_z[s_idx]
+            hij = (h + s_h[s_idx]) * 0.5
+            r = sqrt(xij[0] * xij[0] + xij[1] * xij[1] + xij[2] * xij[2])
+            KERNEL.gradient(xij, r, hij, dwij)
+            dw = sqrt(dwij[0] * dwij[0] + dwij[1] * dwij[1]
+                      + dwij[2] * dwij[2])
+            V = s_m[s_idx] / s_rho[s_idx]
+            if r >= 1.0e-09:
```

Comment:
How did you choose 1e-9?

Reply:
This is to prevent r being 0, changed this to 1.0e-12.